### PR TITLE
Add support for podman push, create, run --retry --retry-delay

### PR DESCRIPTION
**Pull request by rhatdan**
#### Does this PR introduce a user-facing change?

```release-note
Podman  create, push, and run now supports --retry and --retry-delay
```

### CODE REVIEW

**Review by edsantiago — APPROVED** [review]
Tests LGTM

**Comment by mheon** [comment]
Code LGTM

**Comment by TomSweeneyRedHat** [comment]
@rhatdan needs a rebase on this one.
Also, unless I scrolled by them, I didn't see any man page changes and I'd expect them given new options for push, or is it all API?

File: docs/source/markdown/options/retry-delay.md
Comment (TomSweeneyRedHat) [inline docs/source/markdown/options/retry-delay.md:8]:
```suggestion
the registry and local storage in case of failure. The default is to start at two seconds and then exponentially back off. The delay is used when this value is set, and no exponential back off occurs.
```

File: docs/source/markdown/options/retry.md
Comment (TomSweeneyRedHat) [inline docs/source/markdown/options/retry.md:7]:
```suggestion
Number of times to retry pulling or pushing images between the registry and local
```

**Comment by TomSweeneyRedHat** [comment]
Some doc nits, other than that, LGTM once the tests are hip.  Currently, they are redder than red.

**Comment by rhatdan** [comment]
@mheon @Luap99 @TomSweeneyRedHat @edsantiago PTANL

**Comment by edsantiago** [comment]
Tests look adequate; I doubt it's worth the effort to repeat the `pull` test.

Please update the release-note (in comment 0) to reflect create & run.

LGTM if the flakes pass, but needs more review than just mine.

**Comment by rhatdan** [comment]
@ashley-cui @baude @n1hility We are blocking PRs based on containers/common not working with hyperV?

**Comment by mheon** [comment]
Rebase once https://github.com/containers/podman/pull/21572 lands and you should get green tests

**Comment by Luap99** [comment]
@giuseppe PTAL, your latest c/storage changes break podman tests:
```
3f2e81abb38926d9b908528412c25fa2e199b6b9 is the first bad commit
commit 3f2e81abb38926d9b908528412c25fa2e199b6b9
Author: Giuseppe Scrivano <gscrivan@redhat.com>
Date:   Tue Feb 13 13:43:16 2024 +0100

    overlay: replace rmdir with rename
    
    instead of removing the "merged" directory, replace it with a fresh
    empty directory so that we can assume "merged" is always present and
    there is no need to recreate it.
    
    Signed-off-by: Giuseppe Scrivano <gscrivan@redhat.com>

 drivers/overlay/overlay.go | 26 +++++++++++++++++++++-----
 1 file changed, 21 insertions(+), 5 deletions(-)
bisect found first bad commit
$ git bisect log
git bisect start
# status: waiting for both good and bad commits
# bad: [a083950a778fb14c70819d20a42472618f3cc5fe] Merge pull request #1843 from giuseppe/enhance-chunked-cache
git bisect bad a083950a778fb14c70819d20a42472618f3cc5fe
# status: waiting for good commit(s), bad commit known
# good: [387d7c8480b012455286c9804ef920493d9d5140] Merge pull request #1844 from mtrmac/chunked-cleanups
git bisect good 387d7c8480b012455286c9804ef920493d9d5140
# bad: [8a4fb72d2e4560d75ac6dbce4070764985119b29] Merge pull request #1828 from giuseppe/overlay-ignore-mkdir-fail-in-additional-store
git bisect bad 8a4fb72d2e4560d75ac6dbce4070764985119b29
# bad: [de38d80855503aac0c46de79b121588cb4d57966] overlay: do not recreate "merged" dir
git bisect bad de38d80855503aac0c46de79b121588cb4d57966
# bad: [3f2e81abb38926d9b908528412c25fa2e199b6b9] overlay: replace rmdir with rename
git bisect bad 3f2e81abb38926d9b908528412c25fa2e199b6b9
# first bad commit: [3f2e81abb38926d9b908528412c25fa2e199b6b9] overlay: replace rmdir with rename
```

Reproducer: `hack/bats --root 60:"podman mount external container - basic test"`

Upgrade tests are also failing, I haven't run bisect on it and just assume it is the same issue.
```
[+0040s] # [08:43:48.850457652] # podman start -a mydonecontainer
[+0040s] # [08:43:48.918513158] time="2024-02-28T08:43:48-06:00" level=error msg="Unmounting /tmp/pu.Y2o8Z7/root/overlay/df354de822f05c16ec885700f693680f3be79db84467695e40869780a1430cd6/merged: no such file or directory"
[+0040s] # Error: unable to start container 986d1a1d7a84ab5e50d37222b96e3711058d9753f311979e7b0d068437a9ad7f: mounting storage for container 986d1a1d7a84ab5e50d37222b96e3711058d9753f311979e7b0d068437a9ad7f: creating overlay mount to /tmp/pu.Y2o8Z7/root/overlay/df354de822f05c16ec885700f693680f3be79db84467695e40869780a1430cd6/merged, mount_data="lowerdir=/tmp/pu.Y2o8Z7/root/overlay/l/PQ2YAY2HGTEKKRY37OIHRNBFPG:/tmp/pu.Y2o8Z7/root/overlay/l/TWUPN5A2IZF4A7E42PS4DYBTG3,upperdir=/tmp/pu.Y2o8Z7/root/overlay/df354de822f05c16ec885700f693680f3be79db84467695e40869780a1430cd6/diff,workdir=/tmp/pu.Y2o8Z7/root/overlay/df354de822f05c16ec885700f693680f3be79db84467695e40869780a1430cd6/work": no such file or directory
[+0040s] # [08:43:48.924476148] [ rc=125 (** EXPECTED 0 **) ]
```

**Comment by Luap99** [comment]
Reverting 3f2e81abb38926d9b908528412c25fa2e199b6b9 in c/storage seem to fix the test issue here.

**Comment by Luap99** [comment]
Upgrade tests are failing after de38d80855503aac0c46de79b121588cb4d57966, which makes sense because it no longer creates the merged dir thus the mount is failing.

**Comment by rhatdan** [comment]
Once that merges I will rebase this PR and we can get this merged.

**Comment by Luap99** [comment]
please merge https://github.com/containers/storage/pull/1848 then rebase + update c/storage here and then hopefully it passes CI this time around

**Review by Luap99 — APPROVED** [review]
/lgtm

**Pull request merged**
